### ISSUE TITLE
OPeNDAP plug-in issues

### ISSUE THREAD

**Issue by kyang2014** [HDFGroup/hdf-compass#60]
By enabling opendap_model in the compass_viewer, I can build HDF-compass on Linux. So I can test the opendap plug-in at least from my development environment.

In general,
One can open a URL served via DAP via "Open Resource".  A simple netCDF style DAP service can be opened with the following limitations.
1) The current opendap model misses group attributes (at least for root) 

The attributes under the root group(or global attributes) can not be displayed. It seems the feature is missing. 

"
  File "/mnt/scr1/kent/hdf-compass-kent/hdf-compass/build/HDFCompass/out00-PYZ.pyz/compass_viewer.frame", line 388, in on_menu_reopen
  File "/mnt/scr1/kent/hdf-compass-kent/hdf-compass/build/HDFCompass/out00-PYZ.pyz/opendap_model", line 272, in __init__
AttributeError: 'NoneType' object has no attribute 'attributes'

Note: Pydap client  on the command-line can successfully retrieve the global attribute. So this is an issue related to the opendap model. 
Pydap output is as follows:
>>> dataset = open_url('http://test.opendap.org/dap/data/nc/coads_climatology.nc')
>>> print dataset.attributes
{'NC_GLOBAL': {'history': 'FERRET V4.30 (debug/no GUI) 15-Aug-96'}, 'DODS_EXTRA'
: {'Unlimited_Dimension': 'TIME'}}

I think it is promising to fix this issue. It seems the opendap_model just forgets considering the global attribute case.

2) Can not serve any object names that contain special characters

 The opendap plug-in can only address the object names that include underscore,letter and numeric values(CF naming conventions). So essentially all HDF5 data served by the default option of the Hyrax OPeNDAP services can not be served since a "/" is always included in an HDF5 object name. 
Symptom: the plug-in just opens the file without displaying any objects. The file name displayed by the plug-in looks like to be escaped with %...h5 for the dot in an HDF5 file name. 

Pydap command-line client doesn't have an issue to open an HDF5 file served by the default HDF5 handler although I may see %.. displayed for the special characters inside the variable names.

3)  Can and potentially will not serve compound datatype objects well because of the limitation of pydap

My version of the pydap(installed by using pip) can only handle a scalar compound datatype HDF5 dataset. So we should expect the opendap model to be able to serve compound datatype objects well even if bugs are fixed.

**Comment by hyoklee** [comment]
@kyang2014, can you explain how to enable OPeNDAP support in the current compass? I tried 
```
import opendap_model
```

under compass_viewer/__init__.py but nothing happens. Is there a GitHub branch that enabled OPeNDAP?

**Comment by kyang2014** [comment]
That's what I did and it worked. Have you "pyinstaller HDFcompass.spec" from the scratch?

**Comment by giumas** [comment]
I confirm @kyang2014 , I did the same. 
@hyoklee did you try `python HDFCompass.py` and then something like: `http://test.opendap.org/opendap/hyrax/data/hdf5/grid_1_2d.h5` in Open resource file?

**Comment by hyoklee** [comment]
Thanks for reply. I thought it was not working because the URL from the latest Hyrax

http://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5 

did not work.

However, I realized that the following URL works:

http://test.opendap.org/dap/data/nc/coads_climatology.nc

Why doesn't compass work with eosdap? Any clue?

**Comment by giumas** [comment]
This comment is not so clear for me: https://github.com/HDFGroup/hdf-compass/blob/master/compass_viewer/__init__.py#L160

Which issue arised? I have frozen it with `pyinstaller` without troubles.

**Comment by hyoklee** [comment]
@jreadey said enabling opendap screwed Windows packaging.

**Comment by kyang2014** [comment]
What's your definition of not working? I remembered that It works for me if you change http to https.

**Comment by hyoklee** [comment]
![screen shot 2015-10-02 at 11 52 18 am](https://cloud.githubusercontent.com/assets/4994401/10252287/1dd8aa72-68fc-11e5-90a1-a6a6115ff9c5.png)

**Comment by giumas** [comment]
> @jreadey said enabling opendap screwed Windows packaging. 

Most likely, the work done for `3.0.0` of PyInstaller fixed the issue. Incidentally, this part of  https://github.com/HDFGroup/hdf-compass/blob/master/HDFCompass.spec#L29 is not more required.

**Comment by kyang2014** [comment]
weird, Which platform did you try?

**Comment by hyoklee** [comment]
iDAP Mac (Mavericks). I'll try Windows and Linux to see if the error is reproducible.

**Comment by kyang2014** [comment]
Try Linux one that we shared the same environment.

**Comment by giumas** [comment]
On Win7 64bit, only if added the `s` to `http`:

![capture](https://cloud.githubusercontent.com/assets/2849257/10252432/70b3c656-6905-11e5-81c8-383e781afa8d.PNG)

**Comment by jreadey** [comment]
I commented out the opendap plugin import in this commit: c5f075e05b4fa400ec21a3b3e4c05c9c6a63b267.  Anyone who wants to experiement with it will need to uncomment that line!

**Comment by giumas** [comment]
@jreadey, do you remember the issue in the installer? I have not been to replicate it!

**Comment by jreadey** [comment]
@giumas, no, I don't remember.  I'll try it out once I have the builds working for Win & Linux.

**Comment by hyoklee** [comment]
Linux and Windows work file with eosdap.

**Comment by jreadey** [comment]
@kyang2014 - Did you try the running the built executable on another machine?
I created this on Cent0S6: https://s3.amazonaws.com/hdfgroup.download/HDFCompass/latest/linux-centos6_x86_64, but on running on another CentOS box, I got a undefined symbol error: xcb_wait_for_reply64 error on startup.  Doing an install of libxcb.x86_64 resolved the problem.

So I guess pyinstaller isn't picking up all the needed libraries.

**Comment by hyoklee** [comment]
I can run it fine on Platypus.

See the attached screenshot.
[image: Inline image 1]

--
HDF Product Designer: Ideate interoperable Bigdata for Citizen Science and
IoT.

**Comment by hyoklee** [comment]
I learned that I can reach the following OPeNDAP servers with Mac:

[1] http://giraffe:8080/opendap/data/hdf5/grid_1_2d.h5

Success with [1] indicates that The HDF Group's proxy server doesn't work with Mac's PyDAP module. I think it's related to HTTPS redirection and specific to Mac platform only. Update user documentation as a limitation.

**Comment by jreadey** [comment]
@hyoklee: I'm able to open this URL:  http://test.opendap.org/dap/data/nc/coads_climatology.nc.
You can't open it on the HDF Group network?

**Comment by hyoklee** [comment]
Wren machine can't open Eosdap either but can open Giraffe. This behavior is consistent with my iDAP. I don't know what's so special about Mac's network library in handling response from proxy server.

**Comment by hyoklee** [comment]
This is not a good sign because the same https://eosdap URL can be opened by Java tools like Panoply  without any problem. Python network stack (e.g., Pydap or urllib) or Mac's network library underneath is not working properly.

**Comment by jreadey** [comment]
@hyoklee - what's the eosdap URL that's not working?
I'm not able to get a response from http://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5 either from Compass or just doing wget on the command line.

**Comment by hyoklee** [comment]
Try: https://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5 
Please note 's' next to https.

**Comment by jreadey** [comment]
Yes, that works for me.  So doesn't that apply it's HDFGroup network specific?

**Comment by hyoklee** [comment]
Does the URL work in Seattle with HDFCompass? I think I gotta try it at home, later. 
It's so strange that https://eosdap... URL request works on web browsers like Safari but doesn't work for HDFCompass within The HDF Group network.

**Comment by jreadey** [comment]
I actually get an access denied error (403) using this url in Safari: https://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5.  (but it opens in Compass)

**Comment by hyoklee** [comment]
That's wrong URL, drop hyrax in URL. I updated my previous comment.

**Comment by jreadey** [comment]
Yes, that opens in Compass.

**Comment by hyoklee** [comment]
Great! We can say that such HDFCompass problem exists on Macs in The HDF Group network.

**Comment by jreadey** [comment]
Meanwhile on the Windows front, I'm able to open OpenDAP resources when running from the source tree but not from the pyinstaller executable.  Anyone have an idea?  (this is using pyinstaller 3.0)

**Comment by hyoklee** [comment]
I don't have such issue with my Win7 binary created by pyinstaller-2.1.1dev-. Is there a reason to use pyinstaller 3.0?

**Comment by jreadey** [comment]
It will be best to use the same installer for each platform...

**Comment by giumas** [comment]
@jreadey Are you using the 'old' spec file that was crafted for PyInstaller 2.1 with PyInstaller 3.0?

**Comment by giumas** [comment]
Right now, I have been able to open: https://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5 both from source tree and the binary. I have use the spec in the PR https://github.com/HDFGroup/hdf-compass/pull/65

**Comment by jreadey** [comment]
Strange - I'm not able to open the above URL on Windows 7 64-bit  from the source tree (but Mac works).
Here's the console output:
```
C:\Users\John\hdf-compass>python HDFCompass.py
INFO    pydap.request:39 > Opening https://eosdap.hdfgroup.org:8080/opendap/hyra
x/data/hdf5/grid_1_2d.h5.dds
DEBUG   hdf_compass.opendap_model.model.can_handle:56 > able to handle https://e
osdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
DEBUG   hdf_compass.asc_model.model.can_handle:63 > able to handle https://eosda
p.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, not starting with
file://
DEBUG   hdf_compass.bag_model.model.can_handle:72 > able to handle https://eosda
p.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, invalid url
DEBUG   hdf_compass.hdf5_model.model.can_handle:73 > able to handle https://eosd
ap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, not starting with
 file://
DEBUG   hdf_compass.array_model.model.can_handle:78 > able to handle https://eos
dap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
DEBUG   hdf_compass.filesystem_model.model.can_handle:63 > able to handle https:
//eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
INFO    pydap.request:39 > Opening https://eosdap.hdfgroup.org:8080/opendap/hyra
x/data/hdf5/grid_1_2d.h5.dds
DEBUG   hdf_compass.opendap_model.model.can_handle:56 > able to handle https://e
osdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
DEBUG   hdf_compass.asc_model.model.can_handle:63 > able to handle https://eosda
p.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, not starting with
file://
DEBUG   hdf_compass.bag_model.model.can_handle:72 > able to handle https://eosda
p.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, invalid url
DEBUG   hdf_compass.hdf5_model.model.can_handle:73 > able to handle https://eosd
ap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no, not starting with
 file://
DEBUG   hdf_compass.array_model.model.can_handle:78 > able to handle https://eos
dap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
DEBUG   hdf_compass.filesystem_model.model.can_handle:63 > able to handle https:
//eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5? no
```

**Comment by giumas** [comment]
https://eosdap.hdfgroup.org:8080/opendap/hyrax/data/hdf5/grid_1_2d.h5 works on my machine, but a weird error pops up:

![capture](https://cloud.githubusercontent.com/assets/2849257/10713353/5d5e703a-7a88-11e5-931c-e56cbc339645.PNG)

Could it be related to https://github.com/HDFGroup/hdf-compass/pull/50 ?

More in general, I propose to add some debug messages with the releases of the loaded libraries when we load the plugins..

**Comment by giumas** [comment]
After this https://github.com/HDFGroup/hdf-compass/pull/128, I get:

```
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:182 > Python 2.7.10 32bit on Windows 7 (AMD64)
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:184 > numpy 1.10.1
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:185 > matplotlib 1.4.3
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:186 > wxPython 3.0.0.0
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:203 > h5py 2.5.0
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:211 > hydroffice.bag 0.2.3
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:214 > lxml 3.4.4 (libxml 2.9.2, libxslt 1.1.28)
DEBUG   hdf_compass.compass_viewer.viewer.load_plugins:227 > pydap 3.1.1 (protocol 2.15)
```

**Comment by giumas** [comment]
I would suggest to close this one and re-open as multiple tickets.

**Comment by jreadey** [comment]
I'll retest on all platforms later this week and then create issues for any platform specific bugs.

**Comment by giumas** [comment]
It looks like `pydap` 3.2.0 brakes the driver: https://github.com/pydap/pydap/issues/66

**Comment by ax3l** [comment]
This issue is the last one holding back the Python 3 branch #186. Is someone still working on it? :-)

**Comment by ax3l** [comment]
ping @giumas @jreadey opendap is the only HDF-compass backend that does not yet support Python3. Is this going to be continued to be supported and anyone willing to invest time into it? Because it holds back all other plugins migrating to support Python 3 since about a year.

**Comment by jreadey** [comment]
I haven't looked at Compass for a while, but I'm able to run out of the py3 branch and load the opendap plugin though.  @ax3l - is that what you are seeing as well?

I'm not sure how to test it though.  If I do "Open Resource" with http://test.opendap.org/dap/data/nc/coads_climatology.nc, I'm getting exceptions.  Maybe someone more familiar with opendap could weigh in...

**Comment by ax3l** [comment]
Are you running the `py3` branch with Python2 or Python3? With Python 3.6, I do see the following errors in the opendap unit tests:
```
# ...
AttributeError: '<class 'pydap.model.DatasetType'>' object has no attribute 'setdefault'
# ...
```

https://travis-ci.org/ComputationalRadiationPhysics/hdf-compass/builds/451264898

https://github.com/ComputationalRadiationPhysics/hdf-compass/commit/c5bfba51f97cd49b6e4a9dcc1176ff3db4d8797b

**Comment by jreadey** [comment]
Py3 branch with Python3 and yes I get the error too.

Does the error happen with Python2?

**Comment by ax3l** [comment]
I did not check python 2. Did you?
Maybe it's generally outdated.

**Comment by jreadey** [comment]
I tried just now with python 2 but I get this error:

```
  File "hdf_compass/compass_model/__init__.py", line 13, in <module>
    from hdf_compass.compass_model.model import get_stores, push, Store, Node, \
  File "hdf_compass/compass_model/model.py", line 94
    class Store(object, metaclass=ABCMeta):
```

I'm not sure what's going on here...

**Comment by ax3l** [comment]
Can we just disable this plugin and finalize a Python3 release? There are old releases that still ship it and you can re-integrate it with fixes anytime.
